src/err/mod.rs

```diff
@@ -967,6 +967,13 @@ impl PyErrArguments for PyDowncastErrorArguments {
     }
 }
 
+impl<'py, T> std::convert::From<Bound<'py, T>> for PyErr {
```

Comment:
It might be nice to have a bound for `T` here I guess? Even if we come up with some marker trait like `IsPyException` for now.

Seems like the same sort of idea we're discussing in #3879 (where we're talking about what I think we might call `DerefToPyAny`.)

Actually, now that I stare at this, I realise that orphan rules probably allow implementing anything we like for `Bound<$name>`. So we could potentially implement `Into<PyErr> for Bound<$name>` in that `create_exception!` macro.

Then again, I think the worst thing that happens from _not_ adding a bound on `T` here is that non-exception types will create a `SystemError: exceptions must derive from BaseException` (or something like that).

Reply:
Yeah, I was wondering if a bound was necessary, but I saw that everything compiled without one, so I wasn't certain. I'm happy to leave this as is, or to try adding a bound.

Reply:
I might try to sleep on this one and see what I think with fresh eyes...

Reply:
> Actually, now that I stare at this, I realise that orphan rules probably allow implementing anything we like for `Bound<$name>`. So we could potentially implement `Into<PyErr> for Bound<$name>` in that `create_exception!` macro.

Since it's recommended [not to implement `Into`, unless targetting Rust older than 1.41](https://doc.rust-lang.org/std/convert/trait.From.html), I don't think we should try this.

Reply:
I looked into this a bit just now and I can add a trait bound that we implement in the `impl_exception_boilerplate!` macro. What we lose by adding the bound is we can no longer use `.into` when the type is already `Bound<'py, PyAny>` instead of `Bound<'py, Exception>`. We can however still use `PyErr::from_value_bound` directly in this case, so it might be worth it anyway.

Reply:
In terms of compatibility, I'd think it's easier for us to remove the trait bound later than to add it, right? (I.e. give more conversions to users by removing the bound.) So maybe for now if the trait bound lets us match the GIL Ref API closer, it's probably better. We could leave a TODO in the code to note the possibility of relaxing this later?

The conversion from `Bound<'py, PyAny>` is certainly a tempting idea for the future...